core/embed/sec/tropic/tropic.c

```diff
@@ -269,6 +284,231 @@ bool tropic_random_buffer(void *buffer, size_t length) {
+    tropic_ui_progress_t ui_progress,
+    const uint8_t kek[TROPIC_MAC_AND_DESTROY_SIZE],
+    const uint8_t stretched_pins[PIN_MAX_TRIES][TROPIC_MAC_AND_DESTROY_SIZE]) {
+  tropic_driver_t *drv = &g_tropic_driver;
+  lt_ret_t ret = LT_FAIL;
+
+  uint8_t masks[PIN_MAX_TRIES * TROPIC_MAC_AND_DESTROY_SIZE] = {0};
+  for (int i = 0; i < PIN_MAX_TRIES; i++) {
+    for (int j = 0; j < TROPIC_MAC_AND_DESTROY_SIZE; j++) {
+      masks[i * TROPIC_MAC_AND_DESTROY_SIZE + j] =
+          kek[j] ^ stretched_pins[i][j];
+    }
+  }
+
+  ui_progress();
+
+  uint16_t masked_kek_slot =
+      drv->pairing_key_index == TROPIC_UNPRIVILEGED_PAIRING_KEY_SLOT
+          ? TROPIC_KEK_MASKS_UNPRIVILEGED_SLOT
+          : TROPIC_KEK_MASKS_PRIVILEGED_SLOT;
+
+  ret = lt_r_mem_data_erase(&drv->handle, masked_kek_slot);
+  if (ret != LT_OK) {
+    goto cleanup;
+  }
+
+  ui_progress();
+
+  ret = lt_r_mem_data_write(&drv->handle, masked_kek_slot, (uint8_t *)masks,
+                            sizeof(masks));
+  if (ret != LT_OK) {
+    goto cleanup;
+  }
+
+  ui_progress();
+
+cleanup:
+  memzero(masks, sizeof(masks));
+
+  return ret == LT_OK;
+}
+
+bool tropic_unmask_kek(tropic_ui_progress_t ui_progress, uint16_t pin_index,
+                       const uint8_t stretched_pin[TROPIC_MAC_AND_DESTROY_SIZE],
+                       uint8_t kek[TROPIC_MAC_AND_DESTROY_SIZE]) {
+  tropic_driver_t *drv = &g_tropic_driver;
+
+  uint8_t masks[R_MEM_DATA_SIZE_MAX] = {0};
+  _Static_assert(
+      R_MEM_DATA_SIZE_MAX >= PIN_MAX_TRIES * TROPIC_MAC_AND_DESTROY_SIZE,
+      "R_MEM_DATA_SIZE_MAX too small");
+  uint16_t length = 0;
+
+  uint16_t masked_kek_slot =
+      drv->pairing_key_index == TROPIC_UNPRIVILEGED_PAIRING_KEY_SLOT
+          ? TROPIC_KEK_MASKS_UNPRIVILEGED_SLOT
+          : TROPIC_KEK_MASKS_PRIVILEGED_SLOT;
+
+  ui_progress();
+
+  if (lt_r_mem_data_read(&drv->handle, masked_kek_slot, masks, &length) !=
+      LT_OK) {
+    return false;
+  }
+
+  if (length != PIN_MAX_TRIES * TROPIC_MAC_AND_DESTROY_SIZE) {
+    return false;
+  }
+
+  ui_progress();
+
+  for (int i = 0; i < TROPIC_MAC_AND_DESTROY_SIZE; i++) {
+    kek[i] =
+        masks[pin_index * TROPIC_MAC_AND_DESTROY_SIZE + i] ^ stretched_pin[i];
+  }
+  return true;
+}
+
+uint32_t tropic_estimate_time_ms(storage_pin_op_t op, uint16_t pin_index) {
+  const int mac_and_destroy_time_ms = 30;
+  const int pin_set_mac_and_destroy_count = 3 * PIN_MAX_TRIES;
+  const int pin_verify_mac_and_destroy_count = pin_index + 2;
+  const int pin_change_mac_and_destroy_count =
+      pin_set_mac_and_destroy_count + pin_verify_mac_and_destroy_count;
+
+  switch (op) {
+    case STORAGE_PIN_OP_SET:
+      return pin_set_mac_and_destroy_count * mac_and_destroy_time_ms;
+    case STORAGE_PIN_OP_VERIFY:
+      return pin_verify_mac_and_destroy_count * mac_and_destroy_time_ms;
+    case STORAGE_PIN_OP_CHANGE:
+      return pin_change_mac_and_destroy_count * mac_and_destroy_time_ms;
+    default:
+      return 0;
```

Comment:
Ideas for improvement. We could also include the time it takes:
- For SET to erase and write the masks slot, which will be roughly another 30ms.
- For SET to generate the reset key. Not sure whether that should be here though.
- For VERIFY to read the masks slot, probably over 12 ms.